staging/src/k8s.io/api/autoscaling/v2beta1/types.go

```diff
@@ -370,3 +372,203 @@ type HorizontalPodAutoscalerList struct {
+	// +optional
+	Spec VerticalPodAutoscalerSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+	// Current information about the autoscaler.
+	// +optional
+	Status VerticalPodAutoscalerStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// VerticalPodAutoscalerSpec is the specification of the behavior of the autoscaler.
+type VerticalPodAutoscalerSpec struct {
+	// A label query that determines the set of pods controlled by the Autoscaler.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors
+	Selector *metav1.LabelSelector `json:"selector,omitempty" protobuf:"bytes,1,opt,name=selector"`
+
+	// Describes the rules on how changes are applied to the pods.
+	// +optional
+	UpdatePolicy PodUpdatePolicy `json:"updatePolicy,omitempty" protobuf:"bytes,2,opt,name=updatePolicy"`
+
+	// Controls how the autoscaler computes recommended resources.
+	// +optional
+	ResourcePolicy PodResourcePolicy `json:"resourcePolicy,omitempty" protobuf:"bytes,3,opt,name=resourcePolicy"`
+}
+
+// PodUpdatePolicy describes the rules on how changes are applied to the pods.
+type PodUpdatePolicy struct {
+	// Controls when autoscaler applies changes to the pod resoures.
+	// +optional
+	UpdateMode UpdateMode `json:"updateMode,omitempty" protobuf:"bytes,1,opt,name=updateMode"`
+}
+
+// UpdateMode controls when autoscaler applies changes to the pod resoures.
+type UpdateMode string
+
+const (
+	// UpdateModeOff means that autoscaler never changes Pod resources.
+	// The recommender still sets the recommended resources in the
+	// VerticalPodAutoscaler object. This can be used for a "dry run".
+	UpdateModeOff UpdateMode = "Off"
+	// UpdateModeInitial means that autoscaler only assigns resources on pod
+	// creation and does not change them during the lifetime of the pod.
+	UpdateModeInitial UpdateMode = "Initial"
+	// UpdateModeRecreate means that autoscaler assigns resources on pod
+	// creation and additionally can update them during the lifetime of the
+	// pod by deleting and recreating the pod.
+	UpdateModeRecreate UpdateMode = "Recreate"
+	// UpdateModeAuto means that autoscaler assigns resources on pod creation
+	// and additionally can update them during the lifetime of the pod,
+	// using any available update method. Currently this is equivalent to
+	// Recreate, which is the only available update method.
+	UpdateModeAuto UpdateMode = "Auto"
+)
+
+// PodResourcePolicy controls how autoscaler computes the recommended resources
+// for containers belonging to the pod. There can be at most one entry for every
+// named container and optionally a single wildcard entry with Name = "*", which
+// handles all containers that don't have individual policies.
+type PodResourcePolicy struct {
+	// Per-container resource policies.
+	// +optional
+	ContainerPolicies []ContainerResourcePolicy `json:"containerPolicies,omitempty" protobuf:"bytes,1,rep,name=containerPolicies"`
```

Comment:
I think we need `patchStrategy` and `patchMergeKey` on all lists.

@pwittrock

Reply:
Done.